src/main.zig

```diff
@@ -345,6 +357,190 @@ const Node = union(enum) {
+                var node: *BranchNode = root.branch;
+                for (0..depth) |d| {
+                    if (node.commitment == null) {
+                        // consume commitment
+                        node.commitment = commitments[commitment_index];
+                        commitment_index += 1;
+                    }
+
+                    // Add next node in the path if it doesn't exist, unless this is the last one as this is a proof of absence
+                    if (d < depth - 1 and node.children[stem[d]] == .empty) {
+                        node.children[stem[d]].branch = try newEmptyBranchNode(@as(u8, @intCast(d)), crs, alloc);
+                    }
+                }
+
+                // "consume" a state diff
+                statediff_index += 1;
+            },
+            .Other => {
+                const stem = poa_stems[poa_stem_index];
+
+                // Walk the tree depth in order to consume non-initialized commitments
+                var node: *BranchNode = root.branch;
+                for (0..depth) |d| {
+                    if (node.commitment == null) {
+                        commitment_index += 1;
+                    }
+
+                    // Add next node in the path if it doesn't exist, unless this is the last one as this is a proof of absence
+                    if (d < depth - 1 and node.children[stem[d]] == .empty) {
+                        node.children[stem[d]].branch = try newEmptyBranchNode(@as(u8, @intCast(d)), crs, alloc);
+                    }
+                }
+
+                // end of the line: insert a proof of stem marker
+                node.children[stem[depth]].poa_stub.* = stem;
+
+                // "consume" a PoA stem
+                poa_stem_index += 1;
+            },
+            .Present => {
+                const stem = statediffs[statediff_index].stem;
+
+                // Walk the tree depth in order to consume non-initialized commitments
+                var node: *BranchNode = root.branch;
+                for (0..depth) |d| {
+                    if (node.commitment == null) {
+                        commitment_index += 1;
+                    }
+
+                    // Add next node in the path if it doesn't exist, unless this is the last one as this is a proof of absence
+                    if (d < depth - 1 and node.children[stem[d]] == .empty) {
+                        node.children[stem[d]].branch = try newEmptyBranchNode(@as(u8, @intCast(d)), crs, alloc);
+                    }
+                }
+
+                // end of the line: insert a leaf with all present values
+                var ll = try newemptyll(stem, depth, alloc, crs);
+                node.children[stem[depth]] = Node{ .last_level = ll };
+                for (statediffs[statediff_index].suffix_diffs) |suffix_diff| {
+                    // will be left to `null` if suffix_dixx.current_value is `null`.
```

Comment:
note to self: actually this is probably incorrect, given some bugs I've run into